numba/cuda/cudadrv/nvvm.py

```diff
@@ -320,6 +320,11 @@ def find_closest_arch(mycc):
     """
     supported_cc = get_supported_ccs()
 
+    if not supported_cc:
+        msg = "No supported GPU compute capabilities found. " \
+              "Please check your cudatoolkit version matches your CUDA version."
+        raise NvvmSupportError(msg)
```

Comment:
@gmarkall am wondering how this fits with https://github.com/numba/numba/blob/7a864689f45d5258b1ee71d993bba9e114f248e4/numba/cuda/cudadrv/nvvm.py#L291-L292 ? Perhaps this should differentiate between "Runtime not present" and "runtime present but something else is wrong" if that's possible?

Reply:
If there's no toolkit available, the first thing that fails is finding libnvvm, and the traceback and error message look like this:

```
In [1]: from numba import cuda, void, float32

In [2]: @cuda.jit(void(float32))
   ...: def f(x):
   ...:     pass
   ...: 
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/cudadrv/nvvm.py in __new__(cls)
    109                 try:
--> 110                     inst.driver = open_cudalib('nvvm')
    111                 except OSError as e:

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/cudadrv/libs.py in open_cudalib(lib)
     55     path = get_cudalib(lib)
---> 56     return ctypes.CDLL(path)
     57 

~/miniconda3/envs/numbanocuda/lib/python3.8/ctypes/__init__.py in __init__(self, name, mode, handle, use_errno, use_last_error, winmode)
    380         if handle is None:
--> 381             self._handle = _dlopen(self._name, mode)
    382         else:

OSError: libnvvm.so: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

NvvmSupportError                          Traceback (most recent call last)
<ipython-input-2-3a92f115ffe4> in <module>
      1 @cuda.jit(void(float32))
----> 2 def f(x):
      3     pass
      4 

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/decorators.py in kernel_jit(func)
    134             targetoptions['link'] = link
    135             targetoptions['opt'] = opt
--> 136             return Dispatcher(func, sigs, bind=bind, targetoptions=targetoptions)
    137 
    138         def device_jit(func):

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/compiler.py in __init__(self, func, sigs, bind, targetoptions)
    806             if len(sigs) > 1:
    807                 raise TypeError("Only one signature supported at present")
--> 808             self.compile(sigs[0])
    809             self._can_compile = False
    810 

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/compiler.py in compile(self, sig)
    933             self.definitions[(cc, argtypes)] = kernel
    934             if self._bind:
--> 935                 kernel.bind()
    936             self.sigs.append(sig)
    937         return kernel

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/compiler.py in bind(self)
    574         Force binding to current CUDA context
    575         """
--> 576         self._func.get()
    577 
    578     @property

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/compiler.py in get(self)
    444         cufunc = self.cache.get(device.id)
    445         if cufunc is None:
--> 446             ptx = self.ptx.get()
    447 
    448             # Link

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/compiler.py in get(self)
    413         if ptx is None:
    414             arch = nvvm.get_arch_option(*cc)
--> 415             ptx = nvvm.llvm_to_ptx(self.llvmir, arch=arch,
    416                                    **self._extra_options)
    417             self.cache[cc] = ptx

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/cudadrv/nvvm.py in llvm_to_ptx(llvmir, **opts)
    472         })
    473 
--> 474     cu = CompilationUnit()
    475     libdevice = LibDevice(arch=opts.get('arch', 'compute_20'))
    476     # New LLVM generate a shorthand for datalayout that NVVM does not know

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/cudadrv/nvvm.py in __init__(self)
    143 class CompilationUnit(object):
    144     def __init__(self):
--> 145         self.driver = NVVM()
    146         self._handle = nvvm_program()
    147         err = self.driver.nvvmCreateProgram(byref(self._handle))

~/miniconda3/envs/numbanocuda/lib/python3.8/site-packages/numba/cuda/cudadrv/nvvm.py in __new__(cls)
    113                     errmsg = ("libNVVM cannot be found. Do `conda install "
    114                               "cudatoolkit`:\n%s")
--> 115                     raise NvvmSupportError(errmsg % e)
    116 
    117                 # Find & populate functions

NvvmSupportError: libNVVM cannot be found. Do `conda install cudatoolkit`:
libnvvm.so: cannot open shared object file: No such file or directory
```

So I think there's already an appropriate error message for that case.

Reply:
Great, thanks for checking.